### PR TITLE
CA-328130 use usb speed to determine bus

### PR DESCRIPTION
**Pull request by lippirk**
Signed-off-by: lippirk <ben.anson@citrix.com>

### CODE REVIEW

File: xc/device.ml [hunk 1]
```diff
@@ -1748,18 +1748,32 @@ module Vusb = struct
...
-        vusb_controller_plug ~xs ~domid ~driver:usb_controller_driver ~driver_id;
-        Printf.sprintf "%s.0" driver_id;
-      end
+    let usb_bus0 = "usb-bus.0" in
+    let ehci0    = "ehci.0" in
+    let get_bus () =
+      let bus_from_speed =
+        match float_of_string_opt speed with
+        | None -> None
+        | Some speed -> if 0. < speed && speed <= 480. then
```
Comment (psafont) [inline xc/device.ml]: If we envision support for more usb devices, we could turn this into a variant
```
let speed_from_float = function
| 1.5 -> `Low (* 1.0 *)
| 12. -> `Full (* 1.1 *)
| 480. -> `Hi (* 2.0 *)
| 5000. -> `Super (* 3.0 *)
| f when f >= 10000. -> `SuperPlus (* fold all future speeds into 3.1 ?*)
| _ -> Erroneous speed reported
```

Comment (lindig) [reply xc/device.ml]: I think matching floats exactly is not a good idea. I think this needs a sequence of tests using `<=`.

File: xc/device.ml [hunk 2]
```diff
@@ -1748,18 +1748,32 @@ module Vusb = struct
...
+                          Some usb_bus0
+                        else if 480. < speed then
+                          Some ehci0
+                        else
+                          None
+      in
+      let get_bus_from_version () =
+        match String.startswith "1" version with
+        | true  ->   usb_bus0
+        | false -> ( vusb_controller_plug ~xs ~domid ~driver:"usb-ehci" ~driver_id:"ehci";
```
Comment (lindig) [inline xc/device.ml]: '(' .. ')' are not required to group the right-hand side of `->`

Comment (lindig) [reply xc/device.ml]: I probably would have used `sscanf` to parse the version number into major/minor:

```
Scanf.sscanf "1.1" "%d.%d" (fun major _minor -> major);;
```

File: xc/device.ml (hunk 1 above)
Comment (psafont) [reply xc/device.ml]: Float equality may be an issue, it's strange we can only get the speed class as a float: it's defined as an [enum](https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/usb/ch9.h#L1137) in Linux

File: xc/device.ml [hunk 3]
```diff
@@ -1748,18 +1748,46 @@ module Vusb = struct
...
+    let get_bus () =
+      let bus_from_speed =
+        match float_of_string_opt speed with
+        | None -> None
+        | Some speed -> match speed_of_float speed with
+          | `Unknown -> None
+          | `Low     -> Some usb_bus0
+          | `Full    -> Some usb_bus0
+          | `High    -> Some ehci0
+          | `Super   -> raise (Xenopsd_error (Internal_error (Printf.sprintf "vusb_plug: superspeed devices are not supported")))
```
Comment (psafont) [inline xc/device.ml]: It'd be better to use `Unimplemented` rather than `Internal_error`

File: xc/device.ml [hunk 4]
```diff
@@ -1750,38 +1750,40 @@ module Vusb = struct
...
-                             `High   (* v2.0 *)
-                           else if x <= 5000. then
-                             `Super  (* v3.0 *)
-                           else
-                             `Unknown
-    in
     let get_bus () =
+      let usb_bus0 = "usb-bus.0" in
+      let ehci0 () =
+        vusb_controller_plug ~xs ~domid ~driver:"usb-ehci" ~driver_id:"ehci";
```
Comment (psafont) [inline xc/device.ml]: Could this qemu message be sent just inside the `if Qemu.is_running ~xs domid then`?

I'd prefer separation between the preparation without side-effects and the interaction with qemu.

and idea would be to pass around a tuple of the bus and prepare_bus `(string * unit -> unit)` that can be executed afterwards.

File: xc/device.ml [hunk 5]
```diff
@@ -1743,43 +1743,70 @@ module Vusb = struct
...
-        "usb-bus.0"
-      else begin
-        (* Here plug usb controller according to the usb version*)
-        let usb_controller_driver = "usb-ehci" in
-        let driver_id = "ehci" in
-        vusb_controller_plug ~xs ~domid ~driver:usb_controller_driver ~driver_id;
-        Printf.sprintf "%s.0" driver_id;
-      end
+    let get_bus () =
+      let vusb_controller_plug ~xs ~domid ~driver ~driver_id =
```
Comment (edwintorok) [inline xc/device.ml]: You already have `~xs ~domid` etc. from the outer scope, do you need to repeat them here?

File: xc/device.ml [hunk 6]
```diff
@@ -1743,43 +1743,70 @@ module Vusb = struct
...
+                               `Full   (* v1.1 *)
+                             else if x <= 480. then
+                               `High   (* v2.0 *)
+                             else if x <= 5000. then
+                               `Super  (* v3.0 *)
+                             else
+                               `Unknown
+      in
+      let bus_from_speed =
+        match float_of_string_opt speed with
```
Comment (edwintorok) [inline xc/device.ml]: See https://github.com/xapi-project/xen-api/pull/3984#discussion_r358164501 this conversion can be avoided.

**Pull request merged**